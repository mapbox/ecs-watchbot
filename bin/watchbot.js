#!/usr/bin/env node

var fastlog = require('fastlog')('watchbot');
var _ = require('underscore');
var sendNotification = require('../lib/notifications')(process.env.NotificationTopic).send;
var watchbot = require('..');

var required = [
  'Cluster',
  'TaskDefinition',
  'Concurrency',
  'QueueUrl',
  'TaskEventQueueUrl',
  'NotificationTopic',
  'StackName',
  'ExponentialBackoff',
  'LogGroupArn',
  'NotifyAfterRetries',
  'AlarmOnEachFailure'
];

var missing = _.difference(required, Object.keys(process.env));
if (missing.length) {
  var err = new Error('Missing from environment: ' + missing.join(', '));
  fastlog.error(err);
  sendNotification('[watchbot] config error', err.message);
  process.exit(1);
}

/**
 * The main Watchbot loop. This function runs continuously on one or more containers,
 * each of which is responsible for polling SQS and spawning tasks to process
 * messages, while maintaining a predefined task concurrency and reporting any failed
 * processing tasks.
 */
watchbot.main(process.env);
