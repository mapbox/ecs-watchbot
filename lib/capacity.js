var AWS = require('aws-sdk');

module.exports = {};
module.exports.run = run;
module.exports.getClusterArn = getClusterArn;
module.exports.getReservations = getReservations;
module.exports.listInstances = listInstances;
module.exports.calculateRoom = calculateRoom;

function run(argv, callback) {
  /* Confirm that region and stack are provided */
  var usage = 'Usage:   worker-capacity <region> <stack_name>\n';
  usage += 'Example: worker-capacity us-east-1 cats-api-staging';

  if (argv.length !== 2) return callback(usage);
  argv = { region: argv[0], stack: argv[1] };

  /* Run functions */
  getClusterArn(argv, (err, clusterArn) => {
    if (err) return callback(err);
    getReservations(argv, (err, reservations) => {
      if (err) return callback(err);
      listInstances(argv, clusterArn, (err, resources) => {
        if (err) return callback(err);
        var result = {
          capacity: calculateRoom(resources, reservations),
          cluster: clusterArn.match(/arn:aws:ecs:.*:\d*:cluster\/(.*)/)[1],
          stack: argv.stack
        };
        return callback(null, result);
      });
    });
  });
}

function getClusterArn(argv, callback) {
  var cfn = new AWS.CloudFormation({ region: argv.region });

  cfn.describeStacks({ StackName: argv.stack }, (err, res) => {
    if (err) return callback(new Error(err));
    if (!res.Stacks[0] || !res.Stacks[0].Outputs.length) return callback(new Error('Check that the provided region and stack name are correct. You may need to re-create your stack to expose the Outputs property.'));

    var cluster = res.Stacks[0].Outputs.find((o) => { return o.OutputKey === 'ClusterArn'; });
    if (!cluster) return callback(new Error('Recreate your stack to expose the cluster ARN output.'));
    return callback(null, cluster.OutputValue);
  });
}

function getReservations(argv, callback) {
  var cfn = new AWS.CloudFormation({ region: argv.region });
  var ecs = new AWS.ECS({ region: argv.region });

  cfn.describeStackResources({ StackName: argv.stack }, (err, res) => {
    if (err) return callback(new Error(err));
    var worker = res.StackResources.find((r) => { return r.LogicalResourceId === 'WatchbotWorker'; }).PhysicalResourceId;
    ecs.describeTaskDefinition({ taskDefinition: worker }, (err, res) => {
      if (err) return callback(new Error(err));
      var def = res.taskDefinition.containerDefinitions[0];
      return callback(null, {
        Memory: def.memory || def.memoryReservation,
        Cpu: def.cpu
      });
    });
  });
}

function listInstances(argv, cluster, callback) {
  var ecs = new AWS.ECS({ region: argv.region });

  var resources = [];
  ecs.listContainerInstances({ cluster: cluster }).eachPage((err, data, done) => {
    if (err) return callback(new Error(err));
    if (!data) return callback(null, resources);
    ecs.describeContainerInstances({ cluster: cluster, containerInstances: data.containerInstanceArns }, (err, data) => {
      if (err) return callback(new Error(err));
      data.containerInstances.forEach((i) => { resources.push(i.remainingResources); });
      done();
    });
  });
}

function calculateRoom(resources, reservations) {
  return resources.reduce(function(memo, instance) {
    var workerCpu = instance.find((e) => { return e.name === 'CPU'; }).integerValue / reservations.Cpu;
    var workerMemory = instance.find((e) => { return e.name === 'MEMORY'; }).integerValue / reservations.Memory;
    return memo + Number(Math.min(workerCpu, workerMemory).toFixed(0));
  }, 0);
}
