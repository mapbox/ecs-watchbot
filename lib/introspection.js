var http = require('http');
var AWS = require('aws-sdk');

module.exports = introspection;

function introspection(watcherContainerName, cluster, callback) {
  listHostTasks(function(err, data) {
    if (err) return callback(err);

    var watcher = data.Tasks.find(function(task) {
      return task.Containers.find(function(container) {
        return container.Name === watcherContainerName;
      });
    });

    if (!watcher) return callback(new Error('Watcher introspection failed'));

    findServiceId(watcher.Arn, cluster, callback);
  });
}

function listHostTasks(callback) {
  http.get('http://localhost:51678/v1/tasks', function(res) {
    var data = '';
    res
      .on('error', callback)
      .on('data', function(d) { data += d; })
      .on('end', function() { callback(null, JSON.parse(data)); });
  });
}

function findServiceId(arn, cluster, callback) {
  var region = arn.split(':')[3];
  var ecs = new AWS.ECS({ region: region });

  ecs.describeTasks({ tasks: [arn], cluster: cluster }, function(err, data) {
    if (err) return callback(err);
    callback(data.tasks[0].startedBy);
  });
}
