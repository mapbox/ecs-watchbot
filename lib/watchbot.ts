// import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
// import * as sqs from 'aws-cdk-lib/aws-sqs';

export interface CdkProps {
  // Define construct properties here
}

export class Cdk extends Construct {

  constructor(scope: Construct, id: string, props: CdkProps = {}) {
    super(scope, id);

    // Define construct contents here

    // example resource
    // const queue = new sqs.Queue(this, 'CdkQueue', {
    //   visibilityTimeout: cdk.Duration.seconds(300)
    // });
  }
}
