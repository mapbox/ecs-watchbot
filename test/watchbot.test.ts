import {App, Duration, Fn, Stack} from 'aws-cdk-lib';
import { FargateWatchbot, WatchbotProps } from '../lib/watchbot';
import { Template } from 'aws-cdk-lib/assertions';
import {Cluster, ContainerImage} from 'aws-cdk-lib/aws-ecs';
import {Vpc} from "aws-cdk-lib/aws-ec2";
import {CfnLogGroup, RetentionDays} from "aws-cdk-lib/aws-logs";
import {CfnQueue} from "aws-cdk-lib/aws-sqs";
import {CfnTopic} from "aws-cdk-lib/aws-sns";

const stack = new Stack(new App(), 'test-stack', {
    stackName: 'test-stack',
    env: {
        region: 'us-east-1',
        account: '222258372212',
    },
});

const defaultProps = {
    prefix: 'Watchbot',
    containerName: `Watchbot-test-stack`,
    structuredLogging: false,
    writableFileSystem: false,
    maxJobDuration: Duration.seconds(0),
    family: 'serviceName',
    cluster: Cluster.fromClusterAttributes(stack, 'Cluster', {
        clusterName: `fargate-processing-staging`,
        vpc: Vpc.fromLookup(stack, 'VPC', {
            vpcId: 'vpc-id'
        })
    }),
    command: ['watchbot', 'listen'],

    publicIP: false,
    privileged: false,
    logGroupName: `test-stack-us-east-1-watchbot`,
    logGroupRetentionDays: RetentionDays.TWO_WEEKS,
    mountPoints: [{
        containerPath: '/tmp',
        sourceVolume: 'tmp',
        readOnly: true,
    }],
    volumes: [{
        name: 'tmp',
    }],

    fifo: false,
    deadLetterThreshold: 10,
    retentionPeriod: Duration.days(14),
}

describe('FargateWatchbot', () => {
    const requiredProps: WatchbotProps = {
        image: ContainerImage.fromRegistry('amazon/amazon-ecs-sample'),
        serviceName: 'test-service',
        command: ['echo', 'hello'],
        deploymentEnvironment: 'staging',
        serviceVersion: '1.0.0',
    };

    describe('When passing the minimal required props', () => {
        const watchbot = new FargateWatchbot(stack, 'MyFargateWatchbot', requiredProps);

        const template = Template.fromStack(stack);

        it('creates a LogGroup', () => {
            template.hasResourceProperties('AWS::Logs::LogGroup', {
                LogGroupName: defaultProps.logGroupName,
                RetentionInDays: defaultProps.logGroupRetentionDays,
            });
        });

        it('Snap', () => {
            expect(template).toMatchSnapshot()
        })

        it('creates a TaskDefinition', () => {
            template.hasResourceProperties('AWS::ECS::TaskDefinition', {
                Cpu: '256',
                Memory: '512',
                RequiresCompatibilities: ['FARGATE'],
                Family: 'test-service',
                Volumes: defaultProps.volumes.map(v => ({ Name: v.name })),
                ContainerDefinitions: [{
                    Privileged: defaultProps.privileged,
                    Image: 'amazon/amazon-ecs-sample',
                    Command: [...defaultProps.command, ...requiredProps.command],
                    Environment: [
                        { Name: 'QueueUrl', Value: {} },
                        { Name: 'LogGroup', Value: { 'Fn::GetAtt': [ 'WatchbotLogGroup', 'Arn' ]}},
                        { Name: 'writableFilesystem', Value: 'false'},
                        { Name: 'maxJobDuration', Value: '0'},
                        { Name: 'Volumes', Value: '/tmp'},
                        { Name: 'Fifo', Value: 'false'},
                        { Name: 'structuredLogging', Value: 'false'},
                        { Name: 'QUEUE_NAME', Value: {} },
                        { Name: 'WorkTopic', Value: {} },
                    ],
                    LogConfiguration: {
                        LogDriver: 'awslogs',
                        Options: {
                            "awslogs-group": {
                                "Ref": "WatchbotLogGroup",
                            },
                            "awslogs-region": "us-east-1",
                            "awslogs-stream-prefix": requiredProps.serviceVersion,
                        }
                    },
                    MountPoints: [{
                        "ContainerPath": "/tmp",
                        "ReadOnly": true,
                        "SourceVolume": "tmp",
                    }],
                    ReadonlyRootFilesystem: true,
                    Ulimits: [{
                        "HardLimit": 10240,
                        "Name": "nofile",
                        "SoftLimit": 10240,
                    }]
                }]
            });
        });

        it('creates a Fargate service', () => {
            template.hasResourceProperties('AWS::ECS::Service', {
                PropagateTags: "TASK_DEFINITION",
                ServiceName: "test-service",
                Cluster: 'fargate-processing-staging',
                LaunchType: 'FARGATE',
                NetworkConfiguration: {
                    AwsvpcConfiguration: {
                        AssignPublicIp: 'DISABLED'
                    }
                }
            });
        });

        it('creates scaling resources', () => {
            template.hasResourceProperties('AWS::ApplicationAutoScaling::ScalingPolicy', {
                PolicyType: "TargetTrackingScaling",
                TargetTrackingScalingPolicyConfiguration: {
                    "PredefinedMetricSpecification": {
                        "PredefinedMetricType": "ECSServiceAverageCPUUtilization",
                    },
                    "TargetValue": 50,
                },
            });
            template.hasResourceProperties('AWS::ApplicationAutoScaling::ScalingPolicy', {
                "PolicyType": "StepScaling",
                "StepScalingPolicyConfiguration": {
                    "AdjustmentType": "ChangeInCapacity",
                    "MetricAggregationType": "Maximum",
                    "StepAdjustments": [
                        {
                            "MetricIntervalLowerBound": 0,
                            "MetricIntervalUpperBound": 400,
                            "ScalingAdjustment": 1,
                        },
                        {
                            "MetricIntervalLowerBound": 400,
                            "ScalingAdjustment": 5,
                        },
                    ],
                },
            });

            // Scale up
            template.hasResourceProperties('AWS::CloudWatch::Alarm', {
                "AlarmActions": [
                    {}
                ],
                "AlarmDescription": "Upper threshold scaling alarm",
                "ComparisonOperator": "GreaterThanOrEqualToThreshold",
                "Dimensions": [
                    {
                        "Name": "QueueName",
                        "Value": {
                            "Fn::GetAtt": [
                                "MyFargateWatchbotQueueF9A1A651",
                                "QueueName",
                            ],
                        },
                    },
                ],
                "EvaluationPeriods": 1,
                "MetricName": "ApproximateNumberOfMessagesVisible",
                "Namespace": "AWS/SQS",
                "Period": 300,
                "Statistic": "Maximum",
                "Threshold": 100,
            });

            template.hasResourceProperties('AWS::CloudWatch::Alarm', {
                "AlarmActions": [
                    {}
                ],
                "AlarmDescription": "Lower threshold scaling alarm",
                "ComparisonOperator": "LessThanOrEqualToThreshold",
                "Dimensions": [
                    {
                        "Name": "QueueName",
                        "Value": {
                            "Fn::GetAtt": [
                                "MyFargateWatchbotQueueF9A1A651",
                                "QueueName",
                            ],
                        },
                    },
                ],
                "EvaluationPeriods": 1,
                "MetricName": "ApproximateNumberOfMessagesVisible",
                "Namespace": "AWS/SQS",
                "Period": 300,
                "Statistic": "Maximum",
                "Threshold": 0,
            })
        })

        it('creates a 2 SQS queue', () => {
            template.resourceCountIs('AWS::SQS::Queue', 2);

            // Queue
            template.hasResourceProperties('AWS::SQS::Queue', {
                "ContentBasedDeduplication": defaultProps.fifo, // this is only true if fifo is true
                "FifoQueue": defaultProps.fifo,
                "MessageRetentionPeriod": defaultProps.retentionPeriod.toSeconds(),
                "QueueName": "test-stack-WatchbotQueue",
                "RedrivePolicy": {
                    "maxReceiveCount": 10,
                },
                "VisibilityTimeout": 180,
            });

            // DLQ
            template.hasResourceProperties('AWS::SQS::Queue', {
                "ContentBasedDeduplication": defaultProps.fifo, // this is only true if fifo is true
                "FifoQueue": defaultProps.fifo,
                "MessageRetentionPeriod": defaultProps.retentionPeriod.toSeconds(),
                "QueueName": "test-stack-WatchbotDeadLetterQueue",
            });
        });

        it('creates an SNS topic', () => {
            template.hasResourceProperties('AWS::SNS::Topic', {
                TopicName: 'test-stack-WatchbotTopic'
            });

            template.resourceCountIs('AWS::SNS::Subscription', 1);
        });
    });

});
