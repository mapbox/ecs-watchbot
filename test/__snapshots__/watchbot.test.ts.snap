// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FargateWatchbot When passing the minimal required props Snap 1`] = `
{
  "Outputs": {
    "MyFargateWatchbotServiceSQSQueueA736D234": {
      "Value": {
        "Fn::GetAtt": [
          "MyFargateWatchbotQueueF9A1A651",
          "QueueName",
        ],
      },
    },
    "MyFargateWatchbotServiceSQSQueueArn8B33ABBC": {
      "Value": {
        "Fn::GetAtt": [
          "MyFargateWatchbotQueueF9A1A651",
          "Arn",
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "MyFargateWatchbotDeadLetterQueueF490A7BD": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "ContentBasedDeduplication": false,
        "FifoQueue": false,
        "MessageRetentionPeriod": 1209600,
        "QueueName": "test-stack-WatchbotDeadLetterQueue",
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "MyFargateWatchbotQueueF9A1A651": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "ContentBasedDeduplication": false,
        "FifoQueue": false,
        "MessageRetentionPeriod": 1209600,
        "QueueName": "test-stack-WatchbotQueue",
        "RedrivePolicy": {
          "deadLetterTargetArn": {
            "Fn::GetAtt": [
              "MyFargateWatchbotDeadLetterQueueF490A7BD",
              "Arn",
            ],
          },
          "maxReceiveCount": 10,
        },
        "VisibilityTimeout": 180,
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "MyFargateWatchbotQueuePolicy5FF0F7E7": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "sqs:SendMessage",
              "Condition": {
                "ArnEquals": {
                  "aws:SourceArn": {
                    "Ref": "MyFargateWatchbotTopic81B7C8EE",
                  },
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Service": "sns.amazonaws.com",
              },
              "Resource": {
                "Fn::GetAtt": [
                  "MyFargateWatchbotQueueF9A1A651",
                  "Arn",
                ],
              },
            },
            {
              "Action": [
                "sqs:SendMessage",
                "sqs:GetQueueAttributes",
                "sqs:GetQueueUrl",
              ],
              "Condition": {
                "ArnEquals": {
                  "aws:SourceArn": {
                    "Ref": "MyFargateWatchbotTopic81B7C8EE",
                  },
                },
              },
              "Effect": "Allow",
              "Principal": {
                "AWS": "*",
              },
              "Resource": {
                "Fn::GetAtt": [
                  "MyFargateWatchbotQueueF9A1A651",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Queues": [
          {
            "Ref": "MyFargateWatchbotQueueF9A1A651",
          },
        ],
      },
      "Type": "AWS::SQS::QueuePolicy",
    },
    "MyFargateWatchbotQueueteststackMyFargateWatchbotTopic0A7E57440F2D9F88": {
      "DependsOn": [
        "MyFargateWatchbotQueuePolicy5FF0F7E7",
      ],
      "Properties": {
        "Endpoint": {
          "Fn::GetAtt": [
            "MyFargateWatchbotQueueF9A1A651",
            "Arn",
          ],
        },
        "Protocol": "sqs",
        "TopicArn": {
          "Ref": "MyFargateWatchbotTopic81B7C8EE",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "MyFargateWatchbotServiceQueueProcessingFargateService5279DF34": {
      "DependsOn": [
        "MyFargateWatchbotServiceQueueProcessingTaskDefTaskRoleDefaultPolicy04A6B48A",
        "MyFargateWatchbotServiceQueueProcessingTaskDefTaskRoleE1E24202",
      ],
      "Properties": {
        "Cluster": "fargate-processing-staging",
        "DeploymentConfiguration": {
          "Alarms": {
            "AlarmNames": [],
            "Enable": false,
            "Rollback": false,
          },
          "MaximumPercent": 200,
          "MinimumHealthyPercent": 50,
        },
        "EnableECSManagedTags": false,
        "LaunchType": "FARGATE",
        "NetworkConfiguration": {
          "AwsvpcConfiguration": {
            "AssignPublicIp": "DISABLED",
            "SecurityGroups": [
              {
                "Fn::GetAtt": [
                  "MyFargateWatchbotServiceQueueProcessingFargateServiceSecurityGroup49094411",
                  "GroupId",
                ],
              },
            ],
            "Subnets": [
              "p-12345",
              "p-67890",
            ],
          },
        },
        "PropagateTags": "TASK_DEFINITION",
        "ServiceName": "test-service",
        "TaskDefinition": {
          "Ref": "MyFargateWatchbotServiceQueueProcessingTaskDef4E1EFFF9",
        },
      },
      "Type": "AWS::ECS::Service",
    },
    "MyFargateWatchbotServiceQueueProcessingFargateServiceSecurityGroup49094411": {
      "DependsOn": [
        "MyFargateWatchbotServiceQueueProcessingTaskDefTaskRoleDefaultPolicy04A6B48A",
        "MyFargateWatchbotServiceQueueProcessingTaskDefTaskRoleE1E24202",
      ],
      "Properties": {
        "GroupDescription": "test-stack/MyFargateWatchbot/Service/QueueProcessingFargateService/SecurityGroup",
        "SecurityGroupEgress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "VpcId": "vpc-12345",
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "MyFargateWatchbotServiceQueueProcessingFargateServiceTaskCountTargetCpuScaling6546AF11": {
      "DependsOn": [
        "MyFargateWatchbotServiceQueueProcessingTaskDefTaskRoleDefaultPolicy04A6B48A",
        "MyFargateWatchbotServiceQueueProcessingTaskDefTaskRoleE1E24202",
      ],
      "Properties": {
        "PolicyName": "teststackMyFargateWatchbotServiceQueueProcessingFargateServiceTaskCountTargetCpuScaling33C94A45",
        "PolicyType": "TargetTrackingScaling",
        "ScalingTargetId": {
          "Ref": "MyFargateWatchbotServiceQueueProcessingFargateServiceTaskCountTargetF7270170",
        },
        "TargetTrackingScalingPolicyConfiguration": {
          "PredefinedMetricSpecification": {
            "PredefinedMetricType": "ECSServiceAverageCPUUtilization",
          },
          "TargetValue": 50,
        },
      },
      "Type": "AWS::ApplicationAutoScaling::ScalingPolicy",
    },
    "MyFargateWatchbotServiceQueueProcessingFargateServiceTaskCountTargetF7270170": {
      "DependsOn": [
        "MyFargateWatchbotServiceQueueProcessingTaskDefTaskRoleDefaultPolicy04A6B48A",
        "MyFargateWatchbotServiceQueueProcessingTaskDefTaskRoleE1E24202",
      ],
      "Properties": {
        "MaxCapacity": 2,
        "MinCapacity": 1,
        "ResourceId": {
          "Fn::Join": [
            "",
            [
              "service/fargate-processing-staging/",
              {
                "Fn::GetAtt": [
                  "MyFargateWatchbotServiceQueueProcessingFargateService5279DF34",
                  "Name",
                ],
              },
            ],
          ],
        },
        "RoleARN": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":iam::222258372212:role/aws-service-role/ecs.application-autoscaling.amazonaws.com/AWSServiceRoleForApplicationAutoScaling_ECSService",
            ],
          ],
        },
        "ScalableDimension": "ecs:service:DesiredCount",
        "ServiceNamespace": "ecs",
      },
      "Type": "AWS::ApplicationAutoScaling::ScalableTarget",
    },
    "MyFargateWatchbotServiceQueueProcessingFargateServiceTaskCountTargetQueueMessagesVisibleScalingLowerAlarmA455F673": {
      "DependsOn": [
        "MyFargateWatchbotServiceQueueProcessingTaskDefTaskRoleDefaultPolicy04A6B48A",
        "MyFargateWatchbotServiceQueueProcessingTaskDefTaskRoleE1E24202",
      ],
      "Properties": {
        "AlarmActions": [
          {
            "Ref": "MyFargateWatchbotServiceQueueProcessingFargateServiceTaskCountTargetQueueMessagesVisibleScalingLowerPolicy60AB646F",
          },
        ],
        "AlarmDescription": "Lower threshold scaling alarm",
        "ComparisonOperator": "LessThanOrEqualToThreshold",
        "Dimensions": [
          {
            "Name": "QueueName",
            "Value": {
              "Fn::GetAtt": [
                "MyFargateWatchbotQueueF9A1A651",
                "QueueName",
              ],
            },
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "ApproximateNumberOfMessagesVisible",
        "Namespace": "AWS/SQS",
        "Period": 300,
        "Statistic": "Maximum",
        "Threshold": 0,
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "MyFargateWatchbotServiceQueueProcessingFargateServiceTaskCountTargetQueueMessagesVisibleScalingLowerPolicy60AB646F": {
      "DependsOn": [
        "MyFargateWatchbotServiceQueueProcessingTaskDefTaskRoleDefaultPolicy04A6B48A",
        "MyFargateWatchbotServiceQueueProcessingTaskDefTaskRoleE1E24202",
      ],
      "Properties": {
        "PolicyName": "teststackMyFargateWatchbotServiceQueueProcessingFargateServiceTaskCountTargetQueueMessagesVisibleScalingLowerPolicyE0D895F0",
        "PolicyType": "StepScaling",
        "ScalingTargetId": {
          "Ref": "MyFargateWatchbotServiceQueueProcessingFargateServiceTaskCountTargetF7270170",
        },
        "StepScalingPolicyConfiguration": {
          "AdjustmentType": "ChangeInCapacity",
          "MetricAggregationType": "Maximum",
          "StepAdjustments": [
            {
              "MetricIntervalUpperBound": 0,
              "ScalingAdjustment": -1,
            },
          ],
        },
      },
      "Type": "AWS::ApplicationAutoScaling::ScalingPolicy",
    },
    "MyFargateWatchbotServiceQueueProcessingFargateServiceTaskCountTargetQueueMessagesVisibleScalingUpperAlarm3ECE77FA": {
      "DependsOn": [
        "MyFargateWatchbotServiceQueueProcessingTaskDefTaskRoleDefaultPolicy04A6B48A",
        "MyFargateWatchbotServiceQueueProcessingTaskDefTaskRoleE1E24202",
      ],
      "Properties": {
        "AlarmActions": [
          {
            "Ref": "MyFargateWatchbotServiceQueueProcessingFargateServiceTaskCountTargetQueueMessagesVisibleScalingUpperPolicy43B1877A",
          },
        ],
        "AlarmDescription": "Upper threshold scaling alarm",
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": [
          {
            "Name": "QueueName",
            "Value": {
              "Fn::GetAtt": [
                "MyFargateWatchbotQueueF9A1A651",
                "QueueName",
              ],
            },
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "ApproximateNumberOfMessagesVisible",
        "Namespace": "AWS/SQS",
        "Period": 300,
        "Statistic": "Maximum",
        "Threshold": 100,
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "MyFargateWatchbotServiceQueueProcessingFargateServiceTaskCountTargetQueueMessagesVisibleScalingUpperPolicy43B1877A": {
      "DependsOn": [
        "MyFargateWatchbotServiceQueueProcessingTaskDefTaskRoleDefaultPolicy04A6B48A",
        "MyFargateWatchbotServiceQueueProcessingTaskDefTaskRoleE1E24202",
      ],
      "Properties": {
        "PolicyName": "teststackMyFargateWatchbotServiceQueueProcessingFargateServiceTaskCountTargetQueueMessagesVisibleScalingUpperPolicy841E001C",
        "PolicyType": "StepScaling",
        "ScalingTargetId": {
          "Ref": "MyFargateWatchbotServiceQueueProcessingFargateServiceTaskCountTargetF7270170",
        },
        "StepScalingPolicyConfiguration": {
          "AdjustmentType": "ChangeInCapacity",
          "MetricAggregationType": "Maximum",
          "StepAdjustments": [
            {
              "MetricIntervalLowerBound": 0,
              "MetricIntervalUpperBound": 400,
              "ScalingAdjustment": 1,
            },
            {
              "MetricIntervalLowerBound": 400,
              "ScalingAdjustment": 5,
            },
          ],
        },
      },
      "Type": "AWS::ApplicationAutoScaling::ScalingPolicy",
    },
    "MyFargateWatchbotServiceQueueProcessingTaskDef4E1EFFF9": {
      "Properties": {
        "ContainerDefinitions": [
          {
            "Command": [
              "watchbot",
              "listen",
              "echo",
              "hello",
            ],
            "Environment": [
              {
                "Name": "QueueUrl",
                "Value": {
                  "Ref": "MyFargateWatchbotQueueF9A1A651",
                },
              },
              {
                "Name": "LogGroup",
                "Value": {
                  "Fn::GetAtt": [
                    "WatchbotLogGroup",
                    "Arn",
                  ],
                },
              },
              {
                "Name": "writableFilesystem",
                "Value": "false",
              },
              {
                "Name": "maxJobDuration",
                "Value": "0",
              },
              {
                "Name": "Volumes",
                "Value": "/tmp",
              },
              {
                "Name": "Fifo",
                "Value": "false",
              },
              {
                "Name": "structuredLogging",
                "Value": "false",
              },
              {
                "Name": "QUEUE_NAME",
                "Value": {
                  "Fn::GetAtt": [
                    "MyFargateWatchbotQueueF9A1A651",
                    "QueueName",
                  ],
                },
              },
              {
                "Name": "WorkTopic",
                "Value": {
                  "Ref": "MyFargateWatchbotTopic81B7C8EE",
                },
              },
            ],
            "Essential": true,
            "Image": "amazon/amazon-ecs-sample",
            "LogConfiguration": {
              "LogDriver": "awslogs",
              "Options": {
                "awslogs-group": {
                  "Ref": "WatchbotLogGroup",
                },
                "awslogs-region": "us-east-1",
                "awslogs-stream-prefix": "1.0.0",
              },
            },
            "MountPoints": [
              {
                "ContainerPath": "/tmp",
                "ReadOnly": true,
                "SourceVolume": "tmp",
              },
            ],
            "Name": "Watchbot-test-stack",
            "Privileged": false,
            "ReadonlyRootFilesystem": true,
            "Ulimits": [
              {
                "HardLimit": 10240,
                "Name": "nofile",
                "SoftLimit": 10240,
              },
            ],
          },
        ],
        "Cpu": "256",
        "ExecutionRoleArn": {
          "Fn::GetAtt": [
            "MyFargateWatchbotServiceQueueProcessingTaskDefExecutionRole6C7B1C59",
            "Arn",
          ],
        },
        "Family": "test-service",
        "Memory": "512",
        "NetworkMode": "awsvpc",
        "RequiresCompatibilities": [
          "FARGATE",
        ],
        "TaskRoleArn": {
          "Fn::GetAtt": [
            "MyFargateWatchbotServiceQueueProcessingTaskDefTaskRoleE1E24202",
            "Arn",
          ],
        },
        "Volumes": [
          {
            "Name": "tmp",
          },
        ],
      },
      "Type": "AWS::ECS::TaskDefinition",
    },
    "MyFargateWatchbotServiceQueueProcessingTaskDefExecutionRole6C7B1C59": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "ecs-tasks.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "MyFargateWatchbotServiceQueueProcessingTaskDefExecutionRoleDefaultPolicyE99BBDF1": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "WatchbotLogGroup",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "MyFargateWatchbotServiceQueueProcessingTaskDefExecutionRoleDefaultPolicyE99BBDF1",
        "Roles": [
          {
            "Ref": "MyFargateWatchbotServiceQueueProcessingTaskDefExecutionRole6C7B1C59",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "MyFargateWatchbotServiceQueueProcessingTaskDefTaskRoleDefaultPolicy04A6B48A": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "sqs:ReceiveMessage",
                "sqs:ChangeMessageVisibility",
                "sqs:GetQueueUrl",
                "sqs:DeleteMessage",
                "sqs:GetQueueAttributes",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "MyFargateWatchbotQueueF9A1A651",
                  "Arn",
                ],
              },
            },
            {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Resource": {
                "Ref": "MyFargateWatchbotTopic81B7C8EE",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "MyFargateWatchbotServiceQueueProcessingTaskDefTaskRoleDefaultPolicy04A6B48A",
        "Roles": [
          {
            "Ref": "MyFargateWatchbotServiceQueueProcessingTaskDefTaskRoleE1E24202",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "MyFargateWatchbotServiceQueueProcessingTaskDefTaskRoleE1E24202": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "ecs-tasks.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },

  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
